src/cert/ssl_cert_provider.cpp

```diff
@@ -39,26 +39,29 @@ namespace mpl = mp::logging;
 namespace
 {
 constexpr auto kLogCategory = "ssl-cert-provider";
-// utility function for checking return code or raw pointer from openssl C-apis
-// TODO: constrain T to int or raw pointer once C++20 concepts is available
 template <typename T>
-void openssl_check(T result, const std::string& errorMessage)
+concept pointer_like = requires(T t)
+{
+    *t;
+    t == nullptr;
+};
+
+// utility function for checking return code or raw pointer from openssl C-apis
+template <pointer_like T>
+void openssl_check(T&& result, const std::string& errorMessage)
 {
-    // TODO: expand std::is_pointer_v<T> check to cover all smart pointers as well
```

Comment:
Small note: I changed this from `T result` to `T&& result` in response to this TODO, so that we could potentially allow `std::unique_ptr<T>` as an argument. Maybe that doesn't make semantic sense though (since this function would be taking ownership of the `unique_ptr`)...

Reply:
Since `T` is "pointer-like", `sizeof(T)` is probably small, at most `2 * sizeof(void*)` ([`shared_ptr` generally uses two pointers internally](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2003/n1431.htm#Memory-footprint)). Since a reference is also the size of a pointer, we're saving at most 1 pointer by using a reference here.

To keep this simple (we're not even using smart pointers here yet), how about we just pass by value and accept that there'll be a copy? I could add a comment explaining this so that if we decide to change it in the future, people understand the reasoning.